File: 20221213_Tuesday/20221213.go
// Leetcode: 931. Minimum Falling Path Sum.
// https://leetcode.com/problems/minimum-falling-path-sum/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 22 ms, faster than 56.98% of Go online submissions for Minimum Falling Path Sum.
// Memory Usage: 5.5 MB, less than 81.40% of Go online submissions for Minimum Falling Path Sum.
// 2022.12.13 Daily Challenge.
const max = 1000000
var m [][]int
var n int
func minFallingPathSum(matrix [][]int) int {
  m,n = matrix, len(matrix)
  for i:=1; i < n; i += 1{
    for j:=0; j < n; j += 1 {
      a := getVal(i-1,j-1)
      b := getVal(i-1,j)
      c := getVal(i-1,j+1)
      m[i][j] += minThree(a,b,c)
    }
  }
  return minVal(m[n-1])
}
func getVal(i,j int) int {
  if ! (0 <= i && i < n) { return max}
  if ! (0 <= j && j < n) { return max}
  return m[i][j]
}
func minThree(a,b,c int) int{
  ans := a
  if b < ans { ans = b}
  if c < ans { ans = c}
  return ans
}
func minVal(x []int) int{
  ans := x[0]
  for _,v := range x {
    if v < ans { ans = v}
  }
  return ans
}

File: 20221213_Tuesday/20221213.rb
# Leetcode: 931. Minimum Falling Path Sum.
# https://leetcode.com/problems/minimum-falling-path-sum/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 164 ms, faster than 28.57% of Ruby online submissions for Minimum Falling Path Sum.
# Memory Usage: 211.9 MB, less than 71.43% of Ruby online submissions for Minimum Falling Path Sum.
# 2022.12.13 Daily Challenge.
# @param {Integer[][]} matrix
# @return {Integer}
def min_falling_path_sum(matrix)
  @m,@n = matrix, matrix.size
  # Start from second row (index one).
  (1...@n).each do |i|
    (0...@n).each do |j|
      min = get_val(i-1,j-1) 
      tmp = get_val(i-1,j)
      min = tmp if min > tmp 
      tmp = get_val(i-1,j+1)
      min = tmp if min > tmp
      @m[i][j] += min
    end
  end
  @m.last.min
end
def get_val(i,j)
  return Float::INFINITY if !i.between?(0,@n-1)
  return Float::INFINITY if !j.between?(0,@n-1)
  return @m[i][j]
end
